### PR TITLE
refactor: fewer response models

### PR DESCRIPTION
**Pull request by eoaksnes**
## Why is this pull request needed?
 
* Make it easier to work with the API
  * Fewer response models, a common response model for todo item that will be used by the get, get_all, update and create endpoints

* Map entity fields to request and response models
  * Attempt to try to re-use entity code in the backend for use-case request and response models (not sure how good an idea this is, but let's try it as an experiment).
  
Relevant information  
* https://stackoverflow.com/questions/52812337/clean-architecture-why-not-using-the-entity-as-request-model-of-the-use-case  

## What does this pull request change?

* Change response type for the update, to return the updated todo item (instead of success)
* Change response type for add, to return the created todo item (instead of success)
* Added `filter_fields` decorator that maps from entity to response and request models 
  * (first version, should probably be made better if it works good)
* Moved validation logic to entities 
  * (tried to use data class, but ended up using padantic, since it was too much work/hassle keeping core not aware of libraries, which is recommended) 

## Issues related to this change:

Resolved https://github.com/equinor/template-fastapi-react/issues/155

### CODE REVIEW

File: api/src/features/todo/use_cases/delete_todo_by_id.py [hunk 1]
```diff
@@ -6,15 +6,17 @@
...
+    id: str, user_id: str, todo_repository: TodoRepositoryInterface
+) -> DeleteTodoByIdResponseModel:
     todo_item = todo_repository.get(id)
     if todo_item is None:
         raise NotFoundException
     if todo_item.user_id != user_id:
         raise MissingPrivilegeException
     todo_repository.delete(id)
-    return DeleteTodoByIdResponse(success=True)
+    return DeleteTodoByIdResponseModel(success=True)
```
Comment (sebastianvitterso) [inline api/src/features/todo/use_cases/delete_todo_by_id.py:22]: DIdn't you say this shouldn't have any response model? Can just return a 204 (no idea how that looks in FastAPI).

File: api/src/features/todo/use_cases/get_todo_by_id.py [hunk 2]
```diff
@@ -1,26 +1,14 @@
...
+from features.todo.shared_models import TodoItemResponseModel
 
 
-def get_todo_by_id_use_case(id: str, user_id: str, todo_repository: TodoRepositoryInterface) -> GetTodoByIdResponse:
+def get_todo_by_id_use_case(id: str, user_id: str, todo_repository: TodoRepositoryInterface) -> TodoItemResponseModel:
     todo_item = todo_repository.get(id)
     if todo_item.user_id != user_id:
         raise MissingPrivilegeException
-    return GetTodoByIdResponse.from_entity(cast(TodoItem, todo_item))
+    return TodoItemResponseModel.parse_obj(cast(TodoItemResponseModel, todo_item))
```
Comment (sebastianvitterso) [inline api/src/features/todo/use_cases/get_todo_by_id.py:14]: What does the `cast` function _actually_ do here?

**Review by sebastianvitterso — COMMENTED** [review]
Two comments, then I'll approve!

**Comment by sebastianvitterso** [comment]
Also the tests fail!

File: web/src/api/generated/.openapi-generator/FILES [hunk 3]
```diff
@@ -9,12 +9,8 @@ common.ts
 configuration.ts
```
Comment (olavis) [inline web/src/api/generated/.openapi-generator/FILES:9]: ✨ Here's a tip ✨
Check references and implementations of the generated types, to see if they're actually useful. Many of the interfaces in this branch aren't being used at all.

**Review by olavis — CHANGES_REQUESTED** [review]
I've only reviewed the web folder, but I found some "symptoms" in the generated models, that should be resolved in the backend.

File: api/src/features/todo/use_cases/get_todo_by_id.py (hunk 2 above)
Comment (eoaksnes) [reply api/src/features/todo/use_cases/get_todo_by_id.py]: should not be there, fixed

**Comment by eoaksnes** [comment]
Sorry but I force-pushed, but have fixed several things.

How to use API in frontend:

image


image


image

How to create mappings to entities using class Config or as input to decorator (no demo for last):

image

image

So what should we do with the response for delete, should we keep the model or not:

image

**Comment by olavis** [comment]
I believe the points below are important to address in order to solve this. This PR is growing big, has several breaking changes, unresolved comments are outdated due to force-pushing and many important decision points needs conclusion in order to review it properly, like:
* What are the relations between the entities (User and TodoItem)?
* What is the response from DELETE?
* Communication between frontend and backend: what are the web client needs from the API and in which format do they communicate
* What properties should (and should not) the TodoItem type contain in the web client, aka what will be the returned response from the API?

**Review by sebastianvitterso — APPROVED** [review]
Looks good! 
To respond to some of @olavis points:
---

> * What are the relations between the entities (User and TodoItem)?

There is currently no User entity, is there? But each user can have `n` TodoItems, while each TodoItem has one user. (1-n)

> * What is the response from DELETE?

Currently `{ success: boolean }` which works fine.

> * Communication between frontend and backend: what are the web client needs from the API and in which format do they communicate

I feel that currently, the communication works well, and since we have typescript in the frontend, it's not too critical how things are structured (details-wise), because we get autocomplete and type hints everywhere.

> * What properties should (and should not) the TodoItem type contain in the web client, aka what will be the returned response from the API?

Ah this is a discussion for another day!

**Comment by eoaksnes** [comment]
I think things are okay for now:

@olavis please go over any comments again

**Comment by olavis** [comment]
> I think things are okay for now:
> 
> @olavis please go over any comments again

@eoaksnes there's still unresolved comments from my last review, I can take a look again when you've gone through them? I see many of my points still holds up to this update.

File: web/src/api/generated/models/todo-item.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+     * @type {string}
+     * @memberof TodoItem
+     */
+    title?: string;
+    /**
+     * 
+     * @type {boolean}
+     * @memberof TodoItem
+     */
+    is_completed?: boolean;
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply web/src/api/generated/models/todo-item.ts:40]: we can do this, but should be in a separate PR

File: web/src/api/generated/models/update-todo-request.ts [hunk 5]
```diff
@@ -31,7 +31,7 @@ export interface UpdateTodoRequest {
      * @type {boolean}
      * @memberof UpdateTodoRequest
      */
-    is_completed: boolean;
+    is_completed?: boolean;
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply web/src/api/generated/models/update-todo-request.ts]: We can do camel case, but should be in a separate PR.

The UpdateTodoRequest model I think is ok.

File: web/src/api/generated/models/add-todo.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * @interface AddTodo
+ */
+export interface AddTodo {
+    /**
+     * 
+     * @type {string}
+     * @memberof AddTodo
+     */
+    title?: string;
+}
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply web/src/api/generated/models/add-todo.ts]: See below on "Not sure if it was a good idea to change the AddTodoRequestModel to title: str.". 

This made the POST use Query parameters, I did not like it so changed back to add-todo.ts.

File: web/src/api/generated/models/delete-todo-by-id-response-model.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * @interface DeleteTodoByIdResponseModel
+ */
+export interface DeleteTodoByIdResponseModel {
+    /**
+     * 
+     * @type {boolean}
+     * @memberof DeleteTodoByIdResponseModel
+     */
+    success: boolean;
+}
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply web/src/api/generated/models/delete-todo-by-id-response-model.ts]: Called it delete-todo-response, think just responded with success is ok for now, maybe we can in a new PR look at some other way of doing this.

File: web/src/api/generated/models/update-todo.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+     * @memberof UpdateTodo
+     */
+    title?: string;
+    /**
+     * 
+     * @type {boolean}
+     * @memberof UpdateTodo
+     */
+    is_completed?: boolean;
+}
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply web/src/api/generated/models/update-todo.ts]: This is needed for updating the todo.

File: web/src/hooks/useTodos.tsx [hunk 9]
```diff
@@ -29,12 +29,12 @@ const useTodos = (): {
     setLoading(true)
     todoAPI
       .create({
-        addTodoRequest: {
+        addTodo: {
           title: title,
         },
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply web/src/hooks/useTodos.tsx]: this is outdated, and should be better now, see code

File: web/src/hooks/useTodos.tsx [hunk 10]
```diff
@@ -59,14 +59,12 @@ const useTodos = (): {
...
+    const index: number = todos.findIndex((item: TodoItem) => item.id === id)
     const todoItem = todos[index]
     todoAPI
       .updateById({
         id: id,
-        updateTodoRequest: {
+        updateTodo: {
           is_completed: !todoItem.is_completed,
           title: todoItem.title,
         },
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply web/src/hooks/useTodos.tsx]: this is outdated and should be better now, see the changed code

File: web/src/hooks/useTodos.tsx [hunk 11]
```diff
@@ -59,14 +59,12 @@ const useTodos = (): {
...
-      (item: AddTodoResponse) => item.id === id
-    )
+    const index: number = todos.findIndex((item: TodoItem) => item.id === id)
     const todoItem = todos[index]
     todoAPI
       .updateById({
         id: id,
-        updateTodoRequest: {
+        updateTodo: {
           is_completed: !todoItem.is_completed,
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply web/src/hooks/useTodos.tsx]: camelCase = new PR

File: web/src/features/todos/todo-list/TodoItem.tsx [hunk 12]
```diff
@@ -7,10 +7,10 @@ import {
   Tooltip,
 } from '@equinor/eds-core-react'
 import { undo, done, remove_outlined } from '@equinor/eds-icons'
-import { AddTodoResponse } from '../../../api/generated'
+import { TodoItem } from '../../../api/generated'
 
-const TodoItem = (props: {
-  todo: AddTodoResponse
+const TodoItemElement = (props: {
```
[the remark this answers is not in the document]
Comment (eoaksnes) [reply web/src/features/todos/todo-list/TodoItem.tsx]: Good idea, done

File: web/src/hooks/useTodos.tsx [hunk 13]
```diff
@@ -1,17 +1,17 @@
 import { useEffect, useState } from 'react'
 import { AxiosError } from 'axios'
-import { AddTodoResponse, ErrorResponse } from '../api/generated'
+import { TodoItem, ErrorResponse } from '../api/generated'
 import { useTodoAPI } from './useTodoAPI'
 
 const useTodos = (): {
-  todos: AddTodoResponse[]
+  todos: TodoItem[]
```
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/hooks/useTodos.tsx:7]: @sebastianvitterso did you create an issue for this when you resolved the conversation?

File: web/src/api/generated/models/delete-todo-by-id-response-model.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/api/generated/models/delete-todo-by-id-response-model.ts]: @eoaksnes have you created an issue for it?

File: web/src/api/generated/models/todo-item.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/api/generated/models/todo-item.ts]: @eoaksnes @sebastianvitterso have you created the issue?

File: web/src/api/generated/models/update-todo.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/api/generated/models/update-todo.ts]: I disagree. @sebastianvitterso how is this resolved?

File: web/src/hooks/useTodos.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (sebastianvitterso) [reply web/src/hooks/useTodos.tsx:7]: Not until now: #159.

File: web/src/api/generated/models/delete-todo-by-id-response.ts [hunk 14]
```diff
@@ -23,7 +23,7 @@ export interface DeleteTodoByIdResponse {
     /**
      * 
      * @type {boolean}
-     * @memberof DeleteTodoByIdResponse
+     * @memberof DeleteTodoByIdResponseModel
      */
     success: boolean;
```
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/api/generated/models/delete-todo-by-id-response.ts]: @sebastianvitterso why is this resolved? Does the duplicate still exist?

File: web/src/hooks/useTodos.tsx [hunk 15]
```diff
@@ -59,17 +55,12 @@ const useTodos = (): {
...
     const todoItem = todos[index]
     todoAPI
-      .updateById({
-        id: id,
-        updateTodoRequest: {
-          is_completed: !todoItem.is_completed,
-          title: todoItem.title,
-        },
+      .updateById(id, {
+        is_completed: !todoItem.is_completed,
```
Comment (olavis) [inline web/src/hooks/useTodos.tsx:62]: ⚠️ `is_completed` is optional, meaning you are "lying" to the compiler here by telling it that the property cannot be null or undefined.

File: web/src/hooks/useTodos.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/hooks/useTodos.tsx]: I left another comment in the updated code.

File: web/src/hooks/useTodos.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/hooks/useTodos.tsx]: Where's the issue then?

File: web/src/hooks/useTodos.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/hooks/useTodos.tsx]: Good. It's easier to verify the new code if you tried to avoid force-pushing, because the conversation "dies" when you overwrite the history.

File: web/src/api/generated/models/add-todo.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/api/generated/models/add-todo.ts]: Found it. However, it doesn't change the fact that it generates a useless type.. @sebastianvitterso I see you resolved the conversation, have I missed something or is this still present?

File: web/src/api/generated/models/update-todo-request.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (olavis) [reply web/src/api/generated/models/update-todo-request.ts]: @eoaksnes No issue --> no new PR
Also, why do you think this is ok? I've tried to explain why it's problematic, but if you you have some good arguments to override them, I'd really like to hear them!

**Comment by olavis** [comment]
@eoaksnes the second round of reviewing had almost 40 updated files to go through.. Maybe you could focus on avoiding force-pushing and separating changes into smaller PR's next time? 🙏 It's hard to follow your work, and I'm less confident in my reviews as I fear that I easily could've missed something.

**Comment by olavis** [comment]
> @eoaksnes the second round of reviewing had almost 40 updated files to go through.. Maybe you could focus on avoiding force-pushing and separating changes into smaller PR's next time? 🙏 It's hard to follow your work, and I'm less confident in my reviews as I fear that I easily could've missed something.

Okay, had a quick look. I'd say many of my arguments from the first round are still relevant (please explain why you mean they're not worth the attention and I'll appreciate to learn 😸). I also found some new code smells introduced with the update. I think I'll just leave it at this point for now - the changes between each review are big enough to keep this going even longer, and tbh I think it would be more effective to try to move some of this into separate branches and merge them bit by bit, to be able to keep up with the conversations and changes between each round. I see @sebastianvitterso have resolved many of my comments from the first round - if he can vouch for it, good! nothing's better. I think I've come close to the point where I'm not anymore able to review this confidently.

File: web/src/hooks/useTodos.tsx (hunk 15 above)
Comment (sebastianvitterso) [reply web/src/hooks/useTodos.tsx:62]: I might be misunderstanding you, but how is he "lying"? My first assumption is that you're referring to the use of `!` here. In this case, the `!` does not refer to the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator), but simply the negating operator (`!` = `not`).

Comment (olavis) [reply web/src/hooks/useTodos.tsx:62]: You're right! 😅 Well, as I said I'm struggling to focus and keep up with this now haha 😵‍💫 Good that you're still going strong!!
